### PR TITLE
[Bugbash] Bug fix

### PR DESCRIPTION
**Pull request by Bonytu**
#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/contrib/compression/quantization/lsq_quantizer.py [hunk 1]
```diff
@@ -58,10 +63,21 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
         self.evaluator: Evaluator
         self.is_init = False
 
+        self.check_validation()
         self.register_scale()
         self.register_lsq_apply_method()
         self.register_track_func()
 
+    def check_validation(self):
+        for _, ts in self._target_spaces.items():
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/lsq_quantizer.py:72]: `for ts in self._target_spaces.values():`

File: nni/contrib/compression/quantization/dorefa_quantizer.py [hunk 2]
```diff
@@ -41,47 +47,95 @@ class DoReFaQuantizer(Quantizer):
         >>> _, calibration_config = quantizer.compress(max_steps, max_epochs)
     '''
     @overload
-    def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: Evaluator):
+    def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: Evaluator, input_layers: List):
```
Comment (J-shang) [inline nni/contrib/compression/quantization/dorefa_quantizer.py:50]: `input_module_names: List[str] | None  = None`?

File: nni/contrib/compression/quantization/dorefa_quantizer.py [hunk 3]
```diff
@@ -41,47 +47,95 @@ class DoReFaQuantizer(Quantizer):
...
 
     @overload
     def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: Evaluator,
-                 existed_wrappers: Dict[str, ModuleWrapper] | None = None):
+                 input_layers: List=[], existed_wrappers: Dict[str, ModuleWrapper] | None = None):
         ...
 
     def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: Evaluator, \
-                 existed_wrappers: Dict[str, ModuleWrapper] | None = None):
+                 input_layers: List=[], existed_wrappers: Dict[str, ModuleWrapper] | None = None):
```
Comment (J-shang) [inline nni/contrib/compression/quantization/dorefa_quantizer.py:59]: please don't use a obj as function default value

File: nni/contrib/compression/quantization/lsq_quantizer.py [hunk 4]
```diff
@@ -58,10 +63,21 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
...
+            for _, target_space in ts.items():
+                if target_space.quant_scheme != 'symmetric':
```
Comment (J-shang) [inline nni/contrib/compression/quantization/lsq_quantizer.py:74]: what if `quant_scheme` is None

File: nni/contrib/compression/quantization/lsq_quantizer.py (hunk 1 above)
Comment (Bonytu) [reply nni/contrib/compression/quantization/lsq_quantizer.py:72]: done

File: nni/common/version.py [hunk 5]
```diff
@@ -81,3 +81,12 @@ def version_check(expect: dict, raise_error: bool = False) -> None:
                 raise RuntimeError('Version check failed: ' + err_message)
             else:
                 warnings.warn('Version check with warning: ' + err_message)
+
+
+def check_torch_version():
```
Comment (J-shang) [inline nni/common/version.py:86]: function name is not clear, maybe `torch_version_is_2`?

File: nni/contrib/compression/base/apply_method.py [hunk 6]
```diff
@@ -25,39 +25,21 @@ def round_pass(x):
 
     qmax: int = target_space.qmax
     qmin: int = target_space.qmin
+    if target_space._scaler is not None:
+        assert target_space.scale is not None
```
Comment (J-shang) [inline nni/contrib/compression/base/apply_method.py:29]: no need to assert?

File: nni/contrib/compression/base/apply_method.py [hunk 7]
```diff
@@ -95,9 +77,18 @@ def backward(ctx: Any, grad_output: Any) -> Any:
 class ClampRound(torch.autograd.Function):
     @staticmethod
     def forward(ctx: Any, target: torch.Tensor, target_space: QuantizationTargetSpace) -> Any:
-        transformed_target = target_space.zero_point + target / target_space.scale
+        if target_space._scaler is not None:
+            assert target_space.zero_point is not None
```
Comment (J-shang) [inline nni/contrib/compression/base/apply_method.py:81]: can we allow a quantization algo only has scale and zp is None?

File: nni/common/version.py (hunk 5 above)
Comment (Bonytu) [reply nni/common/version.py:86]: good suggestion, I will modify it.

File: nni/contrib/compression/quantization/dorefa_quantizer.py [hunk 8]
```diff
@@ -2,15 +2,29 @@
...
+from nni.common.version import torch_version_is_2
+
 from ..base.compressor import Quantizer
 from ..base.wrapper import ModuleWrapper
 from ..utils.evaluator import Evaluator
-from ..base.target_space import TargetType
+from ..base.target_space import TargetType, QuantizationTargetSpace
+
+ACTIVATION_LIST = [
+    nn.ReLU, nn.RReLU, nn.LeakyReLU, nn.PReLU, nn.Softplus, nn.ELU, nn.CELU, nn.SELU, nn.GELU, \
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/dorefa_quantizer.py:20]: The trailing backslash is unnecessary.

File: nni/contrib/compression/quantization/dorefa_quantizer.py [hunk 9]
```diff
@@ -54,39 +70,129 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
...
+        self.check_validation()
         self.register_dorefa_apply_method()
         self.register_track_func()
 
+    def check_validation(self):
+        for _, ts in self._target_spaces.items():
+            for _, target_space in ts.items():
+                assert target_space.quant_scheme != None
+                if target_space.type is TargetType.PARAMETER and target_space.quant_scheme != 'affine':
+                    warn_msg = f"Only supports affine mode for weight quantization, bug got {target_space.quant_scheme}"
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/dorefa_quantizer.py:82]: NNI prefers single quotes in general. Please align the style.

File: nni/contrib/compression/quantization/dorefa_quantizer.py [hunk 10]
```diff
@@ -54,39 +70,129 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
         super().__init__(model, config_list, evaluator, existed_wrappers=existed_wrappers)
         self.evaluator: Evaluator
         self.is_init = False
-
+        self.check_validation()
         self.register_dorefa_apply_method()
         self.register_track_func()
 
+    def check_validation(self):
+        for _, ts in self._target_spaces.items():
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/dorefa_quantizer.py:78]: `for ts in self._target_spaces.values():`

File: nni/contrib/compression/quantization/dorefa_quantizer.py [hunk 11]
```diff
@@ -54,39 +70,129 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
...
+                assert target_space.quant_scheme != None
+                if target_space.type is TargetType.PARAMETER and target_space.quant_scheme != 'affine':
+                    warn_msg = f"Only supports affine mode for weight quantization, bug got {target_space.quant_scheme}"
+                    _logger.warning(warn_msg)
+                elif target_space.type is TargetType.OUTPUT:
+                    module = target_space._wrapper.module
+                    # case 1: activation module
+                    # case 2: module with activation fused_modules
+                    fused_modules = target_space._wrapper.fused_modules if target_space._wrapper.fused_modules else []
+                    if not isinstance(module, tuple(ACTIVATION_LIST)) and not (len(fused_modules) > 0 and # type: ignore
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/dorefa_quantizer.py:89]: Python do not recommend `len(list) > 0` ([pep8](https://peps.python.org/pep-0008/))

File: nni/contrib/compression/base/compressor.py [hunk 12]
```diff
@@ -294,6 +294,16 @@ def _register_scalers(self):
         # scalers are used to support different sparse/quant granularity
         register_scalers(self._target_spaces, self._set_default_sparse_granularity)  # type: ignore
 
+    def check_target(self, wrapper: ModuleWrapper, target_name: str):
```
Comment (liuzhe-lz) [inline nni/contrib/compression/base/compressor.py:297]: Missing return type.

File: nni/contrib/compression/quantization/dorefa_quantizer.py [hunk 13]
```diff
@@ -54,39 +70,129 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
         super().__init__(model, config_list, evaluator, existed_wrappers=existed_wrappers)
         self.evaluator: Evaluator
         self.is_init = False
-
+        self.check_validation()
         self.register_dorefa_apply_method()
         self.register_track_func()
 
+    def check_validation(self):
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/dorefa_quantizer.py:77]: Missing return type.

File: nni/contrib/compression/quantization/dorefa_quantizer.py [hunk 14]
```diff
@@ -54,39 +70,129 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
...
+
+        target_space._wrapper.module.register_full_backward_pre_hook(quant_dequant_gradient) # type: ignore
+
+    def register_output_backward_hook(self):
+        for _, ts in self._target_spaces.items():
+            is_output = any([target_space.type is TargetType.OUTPUT for target_space in ts.values()])
+            is_param = any([target_space.type is TargetType.PARAMETER for target_space in ts.values()])
+            if is_param and not is_output:
+                if is_proper_torch_version: # torch version >= 2.0.0
+                    for _, target_space in ts.items():
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/dorefa_quantizer.py:130]: values()

File: nni/common/version.py [hunk 15]
```diff
@@ -81,3 +81,12 @@ def version_check(expect: dict, raise_error: bool = False) -> None:
                 raise RuntimeError('Version check failed: ' + err_message)
             else:
                 warnings.warn('Version check with warning: ' + err_message)
+
+
+def torch_version_is_2():
```
Comment (liuzhe-lz) [inline nni/common/version.py:86]: Return type.

File: nni/contrib/compression/base/apply_method.py [hunk 16]
```diff
@@ -25,50 +25,30 @@ def round_pass(x):
...
         target = target.tanh()
         target = target / (2 * target.abs().max()) + 0.5
         dequantized_target = ClampRound.apply(target, target_space)
 
-        return 2 * dequantized_target - 1
+        return 2 * dequantized_target - 1 # type: ignore
 
     @staticmethod
-    def dorefa_clamp_round_input(target: torch.Tensor, target_space: QuantizationTargetSpace):
+    def dorefa_clamp_round_output(target: torch.Tensor, target_space: QuantizationTargetSpace):
```
Comment (liuzhe-lz) [inline nni/contrib/compression/base/apply_method.py:51]: Return type.

File: nni/contrib/compression/quantization/dorefa_quantizer.py [hunk 17]
```diff
@@ -54,39 +70,129 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
...
+                        any([isinstance(item, tuple(ACTIVATION_LIST)) for item in fused_modules[1:]])): # type: ignore
+                        raise ValueError('Output quantization is only supported for activation function or' + \
+                                          f'activation module fusion, but got {type(module)}')
+                    if target_space.quant_scheme != 'affine':
+                        warn_msg = f'Only supports affine mode for output quantization, bug got {target_space.quant_scheme}'
+                        _logger.warning(warn_msg)
+                if target_space._scaler is not None:
+                    raise ValueError('DoRefa Qauntizer doesn\'t support for granularity, please set it to False')
+
+    def _quant_dequant_gradient_hook(self, target_space: QuantizationTargetSpace):
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/dorefa_quantizer.py:99]: Return type.

File: nni/contrib/compression/quantization/dorefa_quantizer.py [hunk 18]
```diff
@@ -54,39 +70,129 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
...
-                    target_space.apply_method = "dofera_clamp_round_output"
+                    target_space.apply_method = 'dorefa_clamp_round_output'
 
     def register_track_func(self):
         for module_name, _ in self._target_spaces.items():
             wrapper = self._module_wrappers[module_name]
             wrapper.register_track_func(self.initialize_scale_zp)
+            wrapper.register_track_func(self.update_scale_zp)
+
+    def update_scale_zp(self, wrapper: ModuleWrapper, target_name: str, target: Tensor):
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/dorefa_quantizer.py:154]: Return type.

File: nni/contrib/compression/quantization/lsq_quantizer.py [hunk 19]
```diff
@@ -58,22 +63,42 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
         self.evaluator: Evaluator
         self.is_init = False
 
+        self.check_validation()
         self.register_scale()
         self.register_lsq_apply_method()
         self.register_track_func()
 
+    def check_validation(self):
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/lsq_quantizer.py:71]: Return type.

File: nni/contrib/compression/quantization/lsq_quantizer.py [hunk 20]
```diff
@@ -58,22 +63,42 @@ def __init__(self, model: torch.nn.Module, config_list: List[Dict], evaluator: E
...
+                    _logger.warning(warn_msg)
+
     def register_track_func(self):
         for module_name, _ in self._target_spaces.items():
             wrapper = self._module_wrappers[module_name]
             wrapper.register_track_func(self.init_scale)
 
     def init_scale(self, wrapper: ModuleWrapper, target_name: str, target: Tensor):
-        if self.is_init or target_name not in wrapper.quantization_target_spaces:
+        def mean_reduce_func(converted_target: Tensor):
```
Comment (liuzhe-lz) [inline nni/contrib/compression/quantization/lsq_quantizer.py:87]: Return type.

**Review by J-shang — APPROVED** [review]

**Review by liuzhe-lz — APPROVED** [review]

**Pull request merged**